### ISSUE TITLE
Translating menus

### ISSUE THREAD

**Issue by KimchiTea** [pmrowla/pylivemaker#19]
Continuing from the lmlsb edit thread, as this is going off-topic.

@LioMajor can you please share some more details about patching menus?

I've tried the hex editor method and the lmlsb edit method, and with both the English text shows up in the menu, but the function is broken; clicking on it does nothing. I'm not sure what other files need to be edited.

**Comment by pmrowla** [comment]
In `ノベルシステム\■初期化.lsb`, there should be 2 separate `Calc StringToArray()` fields, one corresponding to the `システムメニュー項目名` (`system_menu_item_name`)  array variable, and one for the `システムメニュー項目値` (`system_menu_item_value`) variable. The name array controls the displayed menu text, and the values one is used internally in other system scripts when actually handling each menu option.

Are you translating both arrays right now?

My understanding from taking a quick look at the system scripts is that you only want to be translating the fields for the name array, and not the value array (but I have not actually tested this to make sure).

You can see how menu selection is handled in `ノベルシステム/システムメニュー/オプション選択時.lsb` (`novel_system/system_menu/on_option_selection.lsb`).

```
   0: VarNew val ParamType.Str システムメニュー項目値[IndexOfStr(システムメニュー項目名, @ParamStr[0])] 2
   1: If val == "オプション_読んだ文章を自動的に飛ばす"
   2:     <handle skip_read_text option...>
   5: Elseif val == "オプション_フォント選択"
   6:     <handle font_select option...>
   <more elseif cases...>
```

So as far as I can tell, when you make a menu choice in game the following happens:

1. The (translated) text of whatever `システムメニュー項目名` choice you selected is passed into this script as parameter 0 (`@ParamStr[0]`).
2. `IndexOfStr()` is used to get the index of that text string in the name array (`システムメニュー項目名`)
3. That index is used to get the actual internal LM option string from the value array (`システムメニュー項目値`)
4. The internal option string is compared in a long If/Elseif chain, and when the script finds a case that matches the option value, it does stuff for handling that option.

So basically, I think you only want to translate the name array (`システムメニュー項目名`), which will make the game display the menu how you want (w/translated text), but leave the value array alone so that the option values still match the internal LM values used in other system scripts. (Otherwise you would have to translate the If/Elseif cases in the menu selection script so they match in both places)

But again, I haven't actually verified this in a patched game, so liomajor would probably know more for sure.

**Comment by KimchiTea** [comment]
Oh, that's it, thank you! I was assuming you had to change both arrays based on looking at the lsbs for LiveMaker games originally published in English, but I guess the links between the display text and the underlying values behave differently when you change it in LiveMaker directly versus patching it afterwards. I've verified that ノベルシステム/■初期化.lsb works; will try the other ones mentioned [in this post](https://github.com/pmrowla/pylivemaker/issues/11#issuecomment-585117986).

**Comment by pmrowla** [comment]
I think when you set menu strings in LiveMaker, you have the option of setting both the display string and the underlying "value" string, which then get propagated to all of the compiled system lsb scripts whenever you actually build the final game exe/archive. So in that case, the english value strings would appear in all of the If/Elseif comparisons in the on_menu_selection script, and so on.

But for patching an existing game, it will be a lot less work to just translate only the display strings and leave everything else alone.

**Comment by Stefan311** [comment]
I am now on the same topic, I translated the main menu as described and that worked very well.
Now I want to translate all story related ingame choices. I found code snippets like this:
`  35: VarNew _tmp ParamType.Str  2
  36: VarNew _tmpno ParamType.Int  2
  37: VarNew _tmpid ParamType.Str  2
  38: Calc AddArray(_tmp, "カーテンを開ける")
  39: Calc _tmpno = 0
  40: Calc _tmpid = AddDelimiter("\r\n", _tmpid) ++ _tmpno
  41: Calc AddArray(_tmp, "このままにしておく")
  42: Calc _tmpno = 1
  43: Calc _tmpid = AddDelimiter("\r\n", _tmpid) ++ _tmpno
  44: Calc TrimArray(_tmp)
  45: If ArrayToString(_tmp) != ""
  46:     If 1
  47:         Call ノベルシステム\選択メニュー\■選択実行.lsb:0  1 ArrayToString(_tmp) 0 "C" "C" 0 "" "" 1
  48:         Calc 選択実行中 = 0
  49:         Calc StringToArray(_tmpid, _tmpid)
  50:         If 選択番号 >= 0
  51:             Calc 選択番号 = _tmpid[選択番号]
  52:     Else
  53:         Calc 選択値 = _tmp
  54:         Calc 選択番号 = _tmpno
  55:         Calc 選択実行中 = 0
  56:         Calc 右クリックメニュー一時禁止 = 0
  57:         Call ノベルシステム\メッセージボックス\終了.lsb:0  1 
  58:         Wait GetProp("メッセージボックス", 50) != "" 0 0
  59:     ObjDel "VOICE"
  60: VarDel _tmp
  61: VarDel _tmpno
  62: VarDel _tmpid
  63: Jump 000000F5.lsb:105 選択値 == "カーテンを開ける"
  64: Jump 000000F5.lsb:380 1
  65: Wait 1 0 0
`
At first they are feeding all menu text items into an array (line 40 and 43), at last there are conditional jump operations (line 63). 
I have tried to change 40,43,63 with the lmlsb edit function. That seems working, but I don't want to edit all menu items manually (there are around 90 in this game!).

I decided to implement a kind of automatic extracting, translating and patching.
I have two options to do this:
1. quick and dirty, working only for this one game. I have all I need to do this.
2. properly, so I can share the code. But I have some questions before.
-- Is this generic code, is there similar code in every LiveMaker game? How about different LiveMaker versions?
-- Which export file format to use? I would prefer CSV. Easy to save, easy to load, easy to edit by programs like Excel.
-- Should I include a new program to pipe the textfile through google translate? I have this already, but written in C#. Convert this to python should not be a big deal.

**Comment by pmrowla** [comment]
It's standard for LiveMaker3 at least. Anyone building text based choice menu's in the "normal" way from the LiveMaker gui should end up with this kind of code in the lsb. With longer menus you get more `Calc AddArray` sequences at the beginning and more conditional jumps at the end.

I'm not sure if LiveMaker allows building menus w/image choices instead of text? If you're interested in investigating this it will most likely require experimenting with the LM gui and generating test games to see what code gets output into the lsb's.

**Comment by Stefan311** [comment]
Ah, it's standard, good to know. So I will share the code. I will add as many checks as possible to prevent the program from breaking the game functionality.

I have no LifeMaker to check out the image menu choices. To analyse resulting code would be useless for translation anyway. You have to redraw the images in this case.

**Comment by pmrowla** [comment]
Probably also beneficial to move `cli.lmlsb.scan_menus` functionality into the core API as well (related to #33), and only keep the csv related parts in CLI

**Comment by pmrowla** [comment]
closing this since we have the core api and csv tools for translating text menus, and for dumping the needed info about image menus with `extractmenu --lpm`
